### PR TITLE
feat(server): add paged attention to flash models

### LINKED ISSUE #478: Support PagedAttention
### Feature request

[vLLM](https://github.com/vllm-project/vllm) is fast with efficient management of attention key and value memory with PagedAttention, serving higher throughput than TGI.

### Motivation

Adopting PagedAttention would increase throughput and reduce VRAM usage, especially when using beam search. 

### Your contribution

I don't have plan to work on it right now. This is a backlog to me.

### PR DESCRIPTION
**Pull request by OlivierDehaene**
Closes #478

### CODE REVIEW

File: server/Makefile-vllm [hunk 1]
```diff
@@ -0,0 +1,13 @@
+vllm_commit := d284b831c17f42a8ea63369a06138325f73c4cf9
+
+vllm:
+    # Clone vllm
+	git clone https://github.com/OlivierDehaene/vllm.git
+
+build-vllm: vllm
+	cd vllm && git fetch && git checkout $(flash_att_commit)
```
Comment (Narsil) [inline server/Makefile-vllm:8]:
```suggestion
	cd vllm && git fetch && git checkout $(vllm_commit)
```
I think ?

**Comment by LiuXiaoxuanPKU** [comment]
Hi, I met the following error if I build the project from source.

> File "/home/text-generation-inference/server/text_generation_server/models/__init__.py", line 73, in <module>
    from text_generation_server.models.flash_rw import FlashRWSharded
  File "/home/text-generation-inference/server/text_generation_server/models/flash_rw.py", line 9, in <module>
    from text_generation_server.models.custom_modeling.flash_rw_modeling import (
  File "/home/text-generation-inference/server/text_generation_server/models/custom_modeling/flash_rw_modeling.py", line 13, in <module>
    import vllm_cache_ops
ModuleNotFoundError: No module named 'vllm_cache_ops'

Any ideas for fixing it? Thanks a lot!

**Comment by OlivierDehaene** [comment]
Checkout the last commit and `make install-vllm`

**Comment by fxmarty** [comment]
@OlivierDehaene Is it not possible not to disable it (aka not use it) on purpose?

**Comment by OlivierDehaene** [comment]
Yes why?

**Comment by fxmarty** [comment]
It broke my custom docker install. But agree it just adds complexity and maintenance for not much (just the ability to enable/disable optims, just like quantization).

File: server/text_generation_server/models/causal_lm.py [hunk 2]
```diff
@@ -122,7 +122,7 @@ def from_pb(
         position_ids.masked_fill_(tokenized_inputs["attention_mask"] == 0, 1)
         all_input_ids = tokenized_inputs["input_ids"].T.split(1, dim=1)
 
-        max_tokens = len(inputs) * max_input_length + max_decode_tokens
+        max_tokens = len(inputs) * (max_input_length + max_decode_tokens)
```
Comment (zealchen) [inline server/text_generation_server/models/causal_lm.py:125]: HI, Dehaene. max_decode_tokens is cumulated in the for loop before. Why times len(inputs) here?

**Pull request merged**